### PR TITLE
PCHR-3074: Add the CMS User Permission Interface and related classes

### PR DESCRIPTION
**Pull request by tunbola**
## Overview
This PR adds some generic CMS related classes and interfaces to the HRCore extension that can be used by other extensions.
The class added in this PR is the UserPermission Interface and its related classes

## Before
CMS UserPermission interface and related classes were not present.

## After
The following interface was added in this PR:
- CRM_HRCore_CMSData_UserPermissiontInterface: This interface provides functionality to check if a CMS user has access to certain permissions or not.

The following class was added to implement the Interface
- The CRM_HRCore_CMSData_UserPermission_Drupal class

The following factory class was added:
- The CRM_HRCore_CMSData_UserPermissionFactory.

## Comments
- These classes were added primarily to aid the task being done in https://compucorp.atlassian.net/browse/PCHR-3074

### CODE REVIEW

**Comment by tunbola** [comment]
@davialexandre , Some of the requirements for PCHR-3074 is that User Access menu group will be displayed or not depending on the permissions of the contact being viewed on the contact summary page. The CRM_Core_Permission permission class only allows to check for permissions for the current logged in user and it has no method that allows to check permission for a contact/user ID.
Yes the method will be used by isAdmin method.

File: uk.co.compucorp.civicrm.hrcore/CRM/HRCore/CMSData/UserPermission/Drupal.php [hunk 1]
```diff
@@ -0,0 +1,35 @@
+<?php
+
+use CRM_HRCore_CMSData_UserPermissionInterface as UserPermissionInterface;
+
+/**
+ * Class CRM_HRCore_CMSData_UserPermission_Drupal
+ */
+class CRM_HRCore_CMSData_UserPermission_Drupal implements UserPermissionInterface{
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrcore/CRM/HRCore/CMSData/UserPermission/Drupal.php]: missing space before `{`

File: uk.co.compucorp.civicrm.hrcore/tests/phpunit/CRM/HRCore/CMSData/UserPermission/DrupalTest.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+    $contactData = ['cmsId' => 2];
+    $userPermission = new DrupalUserPermissions();
+    $result = $userPermission->check($contactData, ['sample permission']);
+
+    $this->assertTrue($result);
+  }
+
+  public function testCheckPermissionsReturnsFalseWhenUserDoesNotHaveThePermission() {
+    //setting a user ID of 0 will allow us to test the scenario for when a user
+    //has no permission.
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrcore/tests/phpunit/CRM/HRCore/CMSData/UserPermission/DrupalTest.php]: Please explain why

File: uk.co.compucorp.civicrm.hrcore/tests/phpunit/CRM/HRCore/CMSData/UserPermissionFactoryTest.php [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+    $previousUserFramework = CRM_Core_Config::singleton()->userFramework;
+    //We need to manually set this value because civicrm sets it to some other value
+    //when running in testing environment.
+    $cmsFramework = 'UnrecognizedCMS';
+    CRM_Core_Config::singleton()->userFramework = $cmsFramework;
+    $msg = sprintf('Unrecognized CMS: "%s"', $cmsFramework);
+    $this->setExpectedException('Exception', $msg);
+
+    UserPermissionFactory::create();
+    CRM_Core_Config::singleton()->userFramework = $previousUserFramework;
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrcore/tests/phpunit/CRM/HRCore/CMSData/UserPermissionFactoryTest.php]: I'm not sure, but I think this last line will never be called, as the test will stop when the exception is thrown. Can you please check if that is true?

Comment (tunbola) [reply uk.co.compucorp.civicrm.hrcore/tests/phpunit/CRM/HRCore/CMSData/UserPermissionFactoryTest.php]: It does not get called, I updated the PR to use the setUp and tearDown methods to store and reassign the userFramework variable. I updated it for other tests too.

File: uk.co.compucorp.civicrm.hrcore/tests/phpunit/CRM/HRCore/CMSData/UserPermissionFactoryTest.php [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+use CRM_HRCore_CMSData_UserPermissionFactory as UserPermissionFactory;
+use CRM_HRCore_CMSData_UserPermissionInterface as UserPermissionInterface;
+
+/**
+ * Class CRM_HRCore_CMSData_UserPermissionFactoryTest
+ *
+ * @group headless
+ */
+class UserPermissionFactoryTest extends CRM_HRCore_Test_BaseHeadlessTest {
```
Comment (davialexandre) [inline uk.co.compucorp.civicrm.hrcore/tests/phpunit/CRM/HRCore/CMSData/UserPermissionFactoryTest.php]: the class name should be CRM_HRCore_CMSData_UserPermissionFactoryTest

**Pull request merged**